deeppavlov/pipeline_manager/pipeline_manager.py

```diff
@@ -0,0 +1,424 @@
+        self.available_gpu = get_available_gpus(gpu_select=self.use_gpu, gpu_fraction=self.memory_fraction)
+
+        if len(self.available_gpu) == 0:
+            raise ValueError(f"All selected GPU with numbers: ({set(self.use_gpu)}), are busy.")
+        elif self.use_gpu and len(self.available_gpu) < len(self.use_gpu):
+            print(f"PipelineManagerWarning: 'CUDA_VISIBLE_DEVICES' = ({self.use_gpu}), "
+                  f"but only {self.available_gpu} are available.")
+
+        self.num_workers = len(self.available_gpu)
+
+    @staticmethod
+    def train_pipe(pipe: Dict,
+                   i: int,
+                   observer_: ExperimentObserver,
+                   gpu: Optional[int] = None) -> None:
+        """
+        Start learning single pipeline. Observer write all info in log file.
+
+        Args:
+            pipe: config dict of pipeline
+            i:  number of pipeline
+            observer_: link to observer object
+            gpu: index of available gpu
+
+        """
+        # start pipeline time
+        pipe_start = time.time()
+        # modify project environment
+        if gpu is not None:  # gpu can be equal 0
+            os.environ['CUDA_VISIBLE_DEVICES'] = str(gpu)
+        else:
+            os.environ['CUDA_VISIBLE_DEVICES'] = ''
+
+        # run pipeline train with redirected output flow
+        observer_.pipe_ind = i
+        observer_.pipe_conf = pipe['chainer']['pipe']
+        save_path = observer_.build_pipe_checkpoint_folder(pipe, i)
+        proc_out_path = save_path / f"out_{i + 1}.txt"
+        proc_err_path = save_path / f"err_{i + 1}.txt"
+        with proc_out_path.open("a", encoding='utf8') as out_file, proc_err_path.open("a", encoding='utf8') as err_file:
+            with redirect_stderr(err_file), redirect_stdout(out_file):
+                results = train_evaluate_model_from_config(pipe, to_train=True, to_validate=True)
+
+        # add results and pipe time to log
+        observer_.pipe_time = time.strftime('%H:%M:%S', time.gmtime(time.time() - pipe_start))
+        observer_.pipe_res = results
+        # update logger
+        observer_.update_log()
+
+    def process_pool_executor_generator(self,
+                                        pipes_list: Optional[List] = None,
+                                        observer: Optional[ExperimentObserver] = None,
+                                        gpu: bool = False) -> Generator:
+        """
+        Create generator that returning tuple of args fore self.train_pipe method.
+
+        Args:
+            pipes_list:
+            observer: ExperimentObserver object
+            gpu: boolean trigger, determine to use gpu or not
+
+        """
+        if not pipes_list:
+            pipes_list = self.generated_configs
+        if not observer:
+            observer = self.observer
+
+        for i, pipe_conf in enumerate(pipes_list):
+            if gpu:
+                gpu_ind = self.available_gpu[i % len(self.available_gpu)]
```

Comment:
Please pay attention to this piece of code. The gpu on which the training should start may be occupied by another process. Here I tried to solve this problem using the "gpu_free" function. But maybe you can offer a better solution ?